File: _site/technical/leet/shortestPathInBinaryMatrix.py
# 1091

class Solution:
    def __init__(self):
        self.moves = [
            [-1, -1],
            [-1, 0],
            [-1, 1],
            [0, -1],
            [0, 1],
            [1, 0],
            [1, 1],
            [1, -1]
        ]
        self.seen = set()
        self.n_rows = 0
        self.n_cols = 0
        self.resp = float("inf")

    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        if grid[0][0] != 0:
            return(-1)
        
        self.n_rows = len(grid)
        self.n_cols = len(grid[0])
        que = deque([])
        que.appendleft([0, 0, 1])
        while que:
            curr_size = len(que)
            for i in range(curr_size):
                curr = que.pop()
                row = curr[0]
                col = curr[1]
                num_cells = curr[2]
                if row == self.n_rows - 1 and col == self.n_cols - 1:
                    self.resp = min(self.resp, num_cells)

                if (row, col) in self.seen:
                    continue

                
                self.seen.add((row, col))
                for move in self.moves:
                    new_row = row + move[0]
                    new_col = col + move[1]
                    if self.checkRowCol(new_row, new_col) and grid[new_row][new_col] == 0:
                        que.appendleft([new_row, new_col, num_cells + 1])
        
        return(self.resp if self.resp != float("inf") else -1)


    def checkRowCol(self, row, col):
        return(
            row > -1
            and row < self.n_rows
            and col > -1
            and col < self.n_cols
            and (row, col) not in self.seen
        )

File: lsprangers-blog/docs/technical writing/leet/shortestPathInBinaryMatrix.cpp
#include <vector>;
#include <queue>;
using namespace std;
// 1091

class Solution {
public:

    int n_rows;
    int n_cols;
    vector<vector<int>> this_grid;
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        // O(n^2) time and space to initialize all this
        this_grid = grid;
        vector<vector<int>> dirs = { {-1,-1}, {-1,0}, {-1,1}, {0,-1}, {0,1}, {1,-1}, {1,0}, {1,1} };
        queue<vector<int>> que;        
        n_rows = this_grid.size();
        n_cols = this_grid[0].size();
        vector<vector<bool>> seen(n_rows, vector<bool>(n_cols, false));

        if(n_rows == 0 || this_grid[0][0] == 1) {
            return(-1);
        }
        seen[0][0] = true;
        que.push({0, 0, 1});

        while(!que.empty()) {
            int q_size = que.size();
            for(int i = 0; i < que.size(); i++) {
                vector<int> curr = que.front();
                que.pop();
                int row = curr[0];
                int col = curr[1];
                int steps = curr[2];
                if(row == (n_rows - 1) && col == (n_cols - 1)) {
                    return(steps);
                }
                else {
                    for(vector<int> dir : dirs) {
                        int nextRow = row + dir[0];
                        int nextCol = col + dir[1];
                        if(is_valid(nextRow, nextCol) && !seen[nextRow][nextCol]) {
                            seen[nextRow][nextCol] = true;
                            que.push({nextRow, nextCol, steps + 1});
                        }
                    }
                }

            }
            
        }
        return(-1);
    }


    bool is_valid(int row, int col) {
        return(0 <= row && row < n_rows && 0 <= col && col < n_cols && this_grid[row][col] == 0);
    }
};